tests/test_warm_reboot.py

```diff
@@ -780,3 +821,281 @@ def test_swss_port_state_syncup(dvs, testlog):
+    time.sleep(3)
+    dvs.runcmd("ip neigh flush all")
+    time.sleep(1)
+
+    # check neighbors are gone
+    check_kernel_reachable_neigh_num(dvs, 0)
+
+    # start neighsyncd and restore_neighbors
+    marker = dvs.add_log_marker()
+    pubsub = dvs.SubscribeAsicDbObject("SAI_OBJECT_TYPE_NEIGHBOR_ENTRY")
+    start_neighsyncd(dvs)
+    start_restore_neighbors(dvs)
+
+    # should finish the store within 10 seconds
+    time.sleep(10)
+
+    check_kernel_reachable_v4_neigh_num(dvs, NUM_OF_NEIGHS)
+    check_kernel_reachable_v6_neigh_num(dvs, NUM_OF_NEIGHS)
+
+    # check syslog and sairedis.rec file for activities
+    check_syslog_for_neighbor_entry(dvs, marker, 0, 0, "ipv4")
+    check_syslog_for_neighbor_entry(dvs, marker, 0, 0, "ipv6")
+    (nadd, ndel) = dvs.CountSubscribedObjects(pubsub)
+    assert nadd == 0
+    assert ndel == 0
+
+    # check restore Count
+    swss_app_check_RestoreCount_single(state_db, restore_count, "neighsyncd")
+
+    #
+    # Testcase 3:
+    # Stop neighsyncd, appDB entries should be reserved
+    # flush kernel neigh table to simulate warm reboot
+    # Remove half of ips of servers' interfaces, add new half of ips
+    # start neighsyncd, start restore_neighbors service to restore the neighbor table in kernel
+    # check all new neighbors learned in kernel
+    # no changes should be there in syslog and sairedis.rec
+
+    # get restore_count
+    restore_count = swss_get_RestoreCount(state_db)
+
+    # stop neighsyncd and sairedis.rec
+    stop_neighsyncd(dvs)
+    del_entry_tbl(state_db, "NEIGH_RESTORE_TABLE", "Flags")
+
+    # Del half of the ips and a new half of the ips
+    # note: the first ipv4 can not be deleted only
+    for i in range(8, 8+NUM_INTF):
+        for j in range(NUM_NEIGH_PER_INTF/2):
+            dvs.servers[i].runcmd("ip addr del {}.0.0.{}/24 dev eth0".format(i*4, j+NUM_NEIGH_PER_INTF/2+2))
+            dvs.servers[i].runcmd("ip -6 addr del {}00::{}/64 dev eth0".format(i*4,j+NUM_NEIGH_PER_INTF/2+2))
+            dvs.servers[i].runcmd("ip addr add {}.0.0.{}/24 dev eth0".format(i*4, j+NUM_NEIGH_PER_INTF+2))
+            dvs.servers[i].runcmd("ip -6 addr add {}00::{}/64 dev eth0".format(i*4,j+NUM_NEIGH_PER_INTF+2))
+
+    dvs.runcmd("ip neigh flush all")
+    time.sleep(3)
+
+    # check neighbors are gone
+    check_kernel_reachable_neigh_num(dvs, 0)
+
+    # start neighsyncd and restore_neighbors
+    marker = dvs.add_log_marker()
+    start_neighsyncd(dvs)
+    start_restore_neighbors(dvs)
+
+    # should finish the store within 10 seconds
+    time.sleep(10)
+
+    check_kernel_reachable_v4_neigh_num(dvs, NUM_OF_NEIGHS)
+    check_kernel_reachable_v6_neigh_num(dvs, NUM_OF_NEIGHS)
+
+    # check syslog and sairedis.rec file for activities
+    check_syslog_for_neighbor_entry(dvs, marker, 0, 0, "ipv4")
+    check_syslog_for_neighbor_entry(dvs, marker, 0, 0, "ipv6")
+    (nadd, ndel) = dvs.CountSubscribedObjects(pubsub)
+    assert nadd == 0
+    assert ndel == 0
```

Comment:
If the neighbors had IP changes,  why no add/del activities?  arpping(restore_neighbors) result not reflected to kernel?

Also should the pubsub be activated just before start_neighsyncd(dvs)?  Or you are trying to accumulate the nadd/ndel throughout the whole test_system_warmreboot_neighbor_syncup?

Reply:
We delete some IPs  and added some IPs at server side, the stale neighbors will take time to be aged out, so no delete activities at this point.  We have later test cases to simulate the age activities.
The new added IPs at servers side were not populated to kernel yet, it requires ping or other ways to get them into kernel, that is also done in later test cases.

My understanding about the pubsub is that, when we count the objects, it will read the messages since last count not from when you subscribed. In my case, the table/objects are the same and we only need pubsub once. The test results match this expectation.  

Maybe @lguohan can comment on this. Also, do we need undo subscribe to the table? that does not seem to be available now.